src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,193 @@
+
+static struct instance_type
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
+
+static void
+handle_cockpit_ws_exited (int sigid)
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+              return;
+            }
+        }
+    }
+  /* This can only be reached in case cockpit-ws exited fast and the parent
+   * process did not manage to insert its PID into the array
+   */
+  errx (1, "Could not find the process in socket_to_pid array");
+}
+
+static pid_t
+spawn_cockpit_ws (const char *ws_path, int fd,
+                  const char **cockpit_ws_args)
+{
+  pid_t pid = fork ();
+
+  if (pid == -1)
+    {
+      err (1, "fork() failed");
+    }
+  else if (pid > 0)
+    {
+      return pid;
+    }
+  else
+    {
+      const char *args[MAX_COCKPIT_WS_ARGS + 1];
+      int i = 0, res;
+      char pid_str[10];
+
+      /* pass the socket to ws like systemd activation does, see sd_listen_fds(3) */
+      if (dup2 (fd, SD_LISTEN_FDS_START) < 0)
+        err (1, "dup2() failed");
+
+      setenv ("LISTEN_FDS", "1", 1);
+      res = snprintf (pid_str, sizeof (pid_str),  "%i", getpid ());
+      assert (res < sizeof (pid_str));
+      setenv ("LISTEN_PID", pid_str, 1);
+      setenv ("LISTEN_FDNAMES", instance_types[i].sockname, 1);
+
+      args[0] = ws_path;
+      for (i = 0; cockpit_ws_args[i] != NULL; i++)
+        args[i + 1] = cockpit_ws_args[i];
+      args[i + 1] = NULL;
+
+      execv (ws_path, (char **) args);
+
+      err (1, "spawning cockpit-ws instance failed");
+    }
+}
+
+int
+main (int argc, char *argv[])
+{
+  int res;
+
+  if (argc != 3)
+    errx (1, "Usage: socket-activation-helper $WS_PATH $SOCKETS_DIR");
+
+  const char *ws_path = argv[1];
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    ufds[i].fd = -1;
+
+  /* Set up signal handler for when cockpit-ws has exited */
+  struct sigaction sa;
+  sigemptyset (&sa.sa_mask);
+  sa.sa_flags = 0;
+  sa.sa_handler = handle_cockpit_ws_exited;
+  sigaction (SIGCHLD, &sa, NULL);
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    {
+      struct sockaddr_un addr = {
+        .sun_family = AF_UNIX
+      };
+      char socket_path[50];
+      int listen_fd;
+
+      socket_to_pid[i] = 0;
+
+      res = snprintf (socket_path, sizeof (socket_path), "%s/%s", argv[2], instance_types[i].sockname);
+      assert (res < sizeof (socket_path));
+      strcpy (addr.sun_path, socket_path);
```

Comment:
Thanks for fixing the asprintf and the buffer overflow (I was about to comment when you force pushed). But this indirection is pointless -- just directly snprintf it into addr.sun_path.